### ISSUE TITLE
Database lock-up due to huge transaction rollback

### ISSUE THREAD

**Issue by mattgrayisok** [craftcms/cms#9999]
### Description

We've been seeing a sporadic issue with, so far, 6 sites hosted on Servd in which the database locks up due to a huge transaction roll back. We spoke briefly to Brad on a support@ thread but I wanted to shift the issue over here as it seems to be a rare but general problem across multiple projects.

The first symptom we see is a database running with high CPU and elevated response rates. Slower reads and often fully blocked writes. 

Running `show engine innodb status` shows a transaction rollback containing millions of undo log entries:

`ROLLING BACK 101 lock struct(s), heap size 24696, 11025 row lock(s), undo log entries 2792119`

Upon filesystem investigation it seems as though the transaction ran until it filled the undo log to capacity, at which point mysql aborts the transaction and rolls back. (The roll back itself can take anywhere from 1 hour to 1 month to complete so we respond to this by just trashing and rebuilding the database entirely).

Once the rollback starts there's no sensible way for us to figure out what queries the original transaction contained, but we did manage to catch 2 of the projects while they were still applying the transaction, rather than rolling it back. Both of those were running queries similar to:

```
UPDATE `craft_structureelements` SET `rgt`=`rgt`-0, `dateUpdated`='2021-09-10 18:23:27' WHERE (`rgt` >= 344) AND (`root`=169781)
```

A couple of weird things:

- Notice the `-0`
- Although this and similar queries ran millions of times within the transaction, the dateUpdated was always current, not a fixed value, but it's possible the output I was looking at was expanding a NOW() or something. Not 100% sure if this matters.

On the most recent project to experience this we managed to get some relatively good logs from around the period in which the event occurred. It seems as though a `Pruning extra revisions` task was running and a control panel user saved several Entries at the same time.

Speculating: It feels like there might be a race condition or some sort of logic which triggers the prune task to drop into an infinite loop if the entries it's working on are edited during its processing.

Possibly related to #9905 although we can't fix by just restarting the database - the transaction rollback just resumes which is the expected behaviour.

### **Speculative** Steps to reproduce

1. Trigger a Prune revisions task
2. Edit the entries that the prune task is working on

### Additional info

- Craft version: We've only seen it with 3.7+
- PHP version: 7.4 / 8
- Database driver & version: MariaDB 10.5
- Plugins & versions: Varied

**Comment by brandonkelly** [comment]
This should be fixed for the next release – see https://github.com/craftcms/cms/issues/9977#issuecomment-955603385.

**Comment by brandonkelly** [comment]
Initial results are very promising, so we’ve gone ahead and merged that change in, and released Craft 3.7.19 with it 🚀

**Comment by mattgrayisok** [comment]
Thanks! I'll pass this along to any clients if it occurs again
